client/components/date-range/index.js

```diff
@@ -0,0 +1,526 @@
+			}
+
+			return newState;
+		} );
+	}
+
+	/**
+	 * Converts a moment date to a native JS Date object
+	 * @param  {MomentJSDate} momentDate a momentjs date object to convert
+	 * @return {DATE}            the converted JS Date object
+	 */
+	momentDateToNative( momentDate ) {
+		return this.props.moment.isMoment( momentDate ) ? momentDate.toDate() : momentDate;
+	}
+
+	/**
+	 * Converts a native JS Date object to a MomentJS Date object
+	 * @param  {Date} nativeDate date to be converted
+	 * @return {MomentJSDate}            the converted Date
+	 */
+	nativeDateToMoment( nativeDate ) {
+		return this.props.moment( nativeDate );
+	}
+
+	/**
+	 * Formats a given date to the appropriate format for the
+	 * current locale
+	 * @param  {Date|MomentJSDate} date the date to be converted
+	 * @return {String}      the date as a formatted locale string
+	 */
+	dateToHumanReadable( date ) {
+		return this.props.moment( date ).format( 'L' );
+	}
+
+	/**
+	 * 	Gets the locale appropriate date format (eg: "MM/DD/YYYY")
+	 * @return {String} date format as a string
+	 */
+	getLocaleDateFormat() {
+		return this.props.moment.localeData().longDateFormat( 'L' );
+	}
+
+	/**
+	 * Enforces that given date is within the bounds of the
+	 * range specified
+	 * @param  {Moment} date             momentJS instance
+	 * @param  {Moment|Date} options.dateFrom the start of the date range
+	 * @param  {Moment|Date} options.dateTo   the end of the date range
+	 * @return {Moment}                  the date clamped to be within the range
+	 */
+	clampDateToRange( date, { dateFrom, dateTo } ) {
+		// Ensure endDate is within bounds of firstSelectableDate
+		if ( dateFrom && date.isBefore( dateFrom ) ) {
+			date = dateFrom;
+		}
+
+		if ( dateTo && date.isAfter( dateTo ) ) {
+			date = dateTo;
+		}
+
+		return date;
+	}
+
+	/**
+	 * Builds an appropriate disabledDays prop for DatePicker
+	 * based on firstSelectableDate and lastSelectableDate
+	 * config props
+	 *
+	 * See:
+	 * http://react-day-picker.js.org/api/DayPicker/#disabledDays
+	 * http://react-day-picker.js.org/docs/matching-days
+	 *
+	 * @return {array} configuration to be passed to DatePicker as disabledDays prop
+	 */
+	getDisabledDaysConfig() {
+		const { firstSelectableDate, lastSelectableDate } = this.props;
+
+		let config = {};
+
+		if ( firstSelectableDate ) {
+			config = {
+				...config,
+				before: this.momentDateToNative( firstSelectableDate ), // disable all days before today
+			};
+		}
+
+		if ( lastSelectableDate ) {
+			config = {
+				...config,
+				after: this.momentDateToNative( lastSelectableDate ), // disable all days before today
+			};
+		}
+
+		return [ config ];
+	}
+
+	/**
+	 * Renders the Popover component
+	 * @return {ReactComponent} the Popover component
+	 */
+	renderPopover() {
+		const popoverClassNames = classNames( {
+			'date-range__popover': true,
+			'is-dialog-visible': true, // forces to render when inside a modal. Is there a way to detect this from global state?
```

Comment:
Yep. Same here. The comment was from the original PR where I obviously had also felt uncomfortable.

Is there a guru on such things I could ask?

If not I will have to do some more digging.

Reply:
Hmm... not sure who would be most familiar with dialog and popover currently. I actually wrote the first versions of both, but believe they have been modified a good deal since I last touched them. My implementation was before we had global state, etc.